### ISSUE TITLE
AI code clean-up and functionality changes for 1.15

### ISSUE THREAD

**Issue by mattsc** [wesnoth/wesnoth#3450]
This is a collection of ideas of what could be done to the existing AIs in Wesnoth 1.15.  There's no guarantee that any of this will be done, but time permitting I would like to take care of at least some of it
Please add to the list or post comments as you see fit.	

- [x] All Lua AI code
  - [x] Lua code cleanup, updates and changes
    - [x] Check for proxy unit fields that are now directly accessible -- b3058e8  8af988c 
    - [x] Avoid using __cfg if possible (for speed reasons) -- 8af988c  aa59ac5  f4f2a44 
    - [x] Avoid some wesnoth.* functions when possible (for speed reasons)  --  42b4430  7584354  10463fd  dcadab9 
    - [x] Use math.huge instead of 9e99 or nil for initializing max/min variables  cb423b2
    - [x] Replace all uses of print() with std_print() -- 3f54908 
    - [x] Remove unneeded library inclusions --  8a996c5  1680436 
    - [x] General cleanup and simplifications  --  a686639  83c4ef2  133568e  f4f2a44  80ae232 0e7c09f  a23f06d  35d257f  1040dbe  c2635ab  62625fd  3bfd59f  23f3554  fdc3249  a7d8024  a781139  2591e43  5fe283d 
  - [x] Search for todos (and mentions of 1.12, 1.13, ...) in code  --  5e0df80  e3b2cbf 
  - [x] Check for hard-coded values of modifiable parameters  --  0c57ae1 
- [x] Micro AI functionality changes
  - [x] Simple Attack
    - [x] Use aggression instead of own_value_weight --  f4f2a44 
    - [x] ~Add parameter for emphasizing kills~ decided not do to this for now
    - [x] ~Add parameter for die chance limits~ decided not do to this for now
  - [x] Others:
    - [x] Add [avoid] tag functionality  --  7579d18  abaa2dc  d7418ab  873bf2f 
    - [x] Add ability to use named location as configuration key, in addition to coordinate pair  --  dcf9514 
    - [x] Distinguish between healing locations and villages  --  1856377 
  - [x] ~New MAIs?~  None at this time.
- [x] Other Lua AIs and AI libraries
  - [x] Experimental AI 
    - [x] Switch to external CAs  --  5474785  b92d8e8  35ba820  ec45b53   (also related: bug fix 
b92d8e8)
    - [x] Allow custom enemy types to be specified for recruit evaluation when no enemy units are on the list  --  d6da498 
    - [x] Take healthy trait into account
    - [x] Check whether ExpAI recruiting can be made more efficient  --  e379cf7  c134322 
  - [x] ~Switch other AIs that still use the Lua engine to external CAs~
  - [x] Deprecate old patrol.lua and some ai_helper functions  --  e3b86f5  5c73537 
  - [x] ai_helper.find_best_move: return nil instead of empty table when no move found, in order to be consistent with other functions of this kind.  This breaks backward compatibility.  --  cd8af96 
  - [x] ai_helper.get_closest_enemy: actually return the closest enemy.  Breaks backward compatibility.  --  a831edd 
- [x] Campaigns
  - [x] AToTB new custom AI
    - [x] Rebalance as per scenario author's instructions
    - [x] ~Use Simple Attack MAI?  Would require some work on the MAI~ leave as is; too specialized to make sense to introduce into MAI
  - [x] Redo or remove ai_controller, or customize mainline scenarios?  See also #3668  --  420afdf  23d3139 
- [x] C++ AI code
  - [x] Recalling & recruiting
    - [x] Add unit-specific recall cost (forward port 01194139)  --  e1da42ea
    - [x] Fix units on recall list interfering with recruiting (forward port 1597a2ce)  --  c4ebb1ce
    - [x] ~Change default for recruiting when no worthwhile units for recalling are found and add configuration parameter (allow_recruits=none/any/worth_while)~  not worth the trouble and added complexity, in my opinion
  - [x] Remove 1.9 (and 1.7!) todos from AI code: delete, apply changes or update to 1.15 todos  -- see #3695 and #3764
  - [x] [exclude_units] tag for CAs: allow excluding units from individual CAs. Should be implemented via [filter_own] tag for consistency with existing attacks aspect; see #2349  --  implemented in #4581
  - [x] Change leader-related aspects to work with multiple leaders; see #2350 -- 4fd17e92 926662216a 34e3c9103f
  - [x] Convert E_NOT_REACHED_DESTINATION from an error to a warning; see #2344 -- 16ede103a
  - [x] Remove potential for divide-by-zero for XP needed to advance; see [this forum post](https://forums.wesnoth.org/viewtopic.php?f=21&t=49297) -- c81ecca2
  - [x] Remove attack_depth aspect  -- 0ff1d90  3da48bb  7c34313  f53e3b9  f00f9e3 c0bc8c5 f901a0f 
  - [x] Move default ai_algorithm definition out of C++ code  --  073f70b  fcc22a3 
  - [x] Remove unused functions  --  f8871d9  
  - [x] ~Look into AI recruiting issues with liminal units~ as pointed out in #2664  - not an AI issue, see my comment in #2664
  - [x] Add optional mp_rank=key to [ai] tags  --  c731be5 
  - [x] ~Check for hard-coded values of modifiable parameters~  didn't find anything
  - [x] ~Check whether healing of non-village hexes is taken into account correctly~ it is

**Comment by stevecotton** [comment]
Is the "AToTB new custom AI" only for scenario 2, or will there also be changes to scenario 3? I ask because #3389 is also discussing changes that would need scenario 3 to be rebalanced.

**Comment by mattsc** [comment]
This one is really just about the rebalancing needed because of the AI in scenario 2.

BTW, if somebody else is already rebalancing other parts of AToTB and wants to do this too, please go for it.  I'd be happy for anything on this list that I don't need to do myself.

**Comment by stevecotton** [comment]
If Github had a smiley for "+1, but I'm not volunteering to do the work either" then I'd add it to mattsc's reply.

**Comment by CelticMinstrel** [comment]
For all MicroAIs that support key pairss of the form `abc_x,abc_y` to specify a location, please add the ability to instead use the key `abc` to specify a named location.

Basically it would look something like this:

```
if cfg.abc then loc = wesnoth.special_locations[cfg.abc]
elseif cfg.abc_x and cfg.abc_y then loc = {cfg.abc_x, cffg.abc_y}
else loc = default_loc
end
```

That's probably not precisely working code as it doesn't deal with the case where the named location exists and may not be accurate in the format of its return value, but hopefully it gets the general idea across?

**Comment by mattsc** [comment]
Okay, I added this to the list.  I think this is pretty low priority, but it's easy enough.

**Comment by gfgtdf** [comment]
when i implemented `wesnoth.game_config.combat_experience` i noticed that a lot of ai code always use a hardcoded `8` instead of `wesnoth.game_config.kill_experience`

**Comment by mattsc** [comment]
That's a good point.  There are probably quite a few hard-coded parameters in both the Lua and the C++ code.  I added an action item to each category.

**Comment by mattsc** [comment]
I updated the task list at the top of the issue with the current status and what I still want to do about the Lua AIs.  Let me know if you think something's missing.  You can click 'edited' at the very top to see only the changes.

**Comment by mattsc** [comment]
I'm considering adding some changes to the list that affect balance.  In order of increasing impact:
- Add an optional key `mp_rank=` to the `[ai]` tag which determines the order in which AIs are shown in the MP computer player selection menu.  This can be used for eras, modifications or cores to change which AI is first in the list.  This does not affect the default chosen if nothing is selected.
- Rename the 'Experimental AI' to something that sounds ... less experimental.  Maybe 'enhanced default AI'?
- Make the enhanced default AI the default for MP games.
- I am not sure whether this could also become the default for campaigns.  Knowing what the differences to the current default are (they are rather subtle for the most part), it probably does not make too big a difference, but it might mean that some campaigns need some rebalancing.
- Maybe an alternative could be to add an option to the preferences: play with new/old default AI

Comments?  Opinion?  No worries if you disagree.  I'm just throwing out ideas for now and not doing it would simply mean less work for me!
All but the first item would only go into 1.15, not 1.14.

Also, note to self, add the AI issue from #2664.

**Comment by mattsc** [comment]
I played through the first two scenarios of AToTB on both difficulties to check the balance after adding the custom AI for Muff Toras.  It is supposed to be quite easy on the easier difficulty ("beginner"), and reasonably challenging on the higher difficulty ("challenging").  As far as I can tell, both goals are accomplished, with Muff Toras' two skeleton guards on "beginner" and the two revenants on "challenging".  The difficulty of beating Muff Toras feel more or less in line with the rest of the scenario.

Bottom line: as far as I am concerned, this is taken care of and I ticked off the AToTB task in the list above.

Of course, if somebody else wants to rebalance the scenario later, that's fine by me, but as far as this issue is concerned, I consider this done.

**Comment by mattsc** [comment]
I updated the task list in the first post with my current take on what I think should still be done for 1.15 before closing this issue. There are a few other AI issues that I might or might not look into at some point, but if so, that can be done independently.

**Comment by stevecotton** [comment]
Removing attack_depth still needs a change to WoV S1, have just ping'd @sigurdfdragon for it :grin:  https://r.wesnoth.org/p646557

**Comment by stevecotton** [comment]
The Simple Attack demo scenario still works, but the concept and dialogue is about using it to work around the 1.12 AI's lack of a High XP Attack. In 1.15, commenting out the micro_ai lines in the scenario showed the default AI winning a turn quicker than the scenario-specific AI.

Given that I think all of this is your work @mattsc, congratulations :grinning:

**Comment by sigurdfdragon** [comment]
@stevecotton I haved remove attack_depth from WoV on master, around Sept 8th or so.

**Comment by mattsc** [comment]
@stevecotton Thanks.  😃  And you're right about the dialog and behavior, of course.  Do you have a suggestion for what to replace this with?  It doesn't have to be great, or even make a whole lot of sense, it's just meant as a demonstration how to make the AI do attacks it would otherwise not do (or not do at that time).  If not, I'll try to think of something.  Thanks for pointing this out.

**Comment by mattsc** [comment]
After investigating some more, it turns out that the task called "Remove potential for divide-by-zero for XP needed to advance" is neither a divide-by-zero problem nor, for the most part, an AI issue.  See #4578 for details.  I will fix the part that goes wrong in the High XP Attack candidate action and consider that sufficient for checking off this task.

**Comment by mattsc** [comment]
The remaining tasks were taken care of in #4588.  I updated the list above with the respective commits and am closing this.
